hibernate-core/src/main/java/org/hibernate/tool/schema/internal/ColumnDefinitions.java

```diff
@@ -28,8 +28,13 @@
 class ColumnDefinitions {
 
 	static boolean hasMatchingType(Column column, ColumnInformation columnInformation, Metadata metadata, Dialect dialect) {
+		boolean sqlTypeNameEquals = normalize( stripArgs( column.getSqlType( metadata ) ) ).equals( normalize( columnInformation.getTypeName() ) );
+		if ( column.getTypeName() != null ) {
+			// should not infer type name from java type if @Column(columnDefinition) present
+			return sqlTypeNameEquals;
+		}
 		boolean typesMatch = dialect.equivalentTypes( column.getSqlTypeCode(metadata), columnInformation.getTypeCode() )
-				|| normalize( stripArgs( column.getSqlType( metadata ) ) ).equals( normalize( columnInformation.getTypeName() ) );
+				|| sqlTypeNameEquals;
```

Comment:
I mean after this latest change you are now optimizing for the uncommon case, putting all the string manipulation up front. I have no idea if it's a significant performance hit, most probably not, but I did prefer the previous version.

Reply:
> I mean after this latest change you are now optimizing for the uncommon case, putting all the string manipulation up front. I have no idea if it's a significant performance hit, most probably not, but I did prefer the previous version.

Previous version doesn’t update existing data type if columnDefinition added, I will add tests to cover it tomorrow.

Reply:
```java
Long version;
```
First run, It's expected that `version` column added  as `bigint`. 

Now I want it be `integer` but keep java type as `Long`.
```java
@ColumnDefinition(columnDefinition="integer")
Long version;
```
Second run, It detect `Long` equals to `bigint`, and schema will not be updated, that's the behavior of previous version, I think it's a bug, WDYT?

Reply:
Right, but this is not a way of using `columnDefinition` that we encourage or even really support.

For this we tell you to use `@JdbcTypeCode(INTEGER)`, so that we can:

- generate DDL that is portable across different databases,
- correctly add `not null` constraints and sometimes other things, and 
- perform any appropriate type conversions on the Java side, and use the correct methods when writing to/from JDBC (not a problem in this particular very trivial case, but important in general).

And the issue is that since `columnDefinition` is _not_ a SQL type name, but a fragment of native DDL from one ~10 different dialects of SQL, even just _recognizing_ that all we have here in your case is a type name is a nontrivial problem.

I mean, really it's a failing of JPA itself, which I guess should let `@Column` specify a `columnType` from the range of well-defined JDBC types, similar to what our `@JdbcTypeCode` does. Back in the day this was not added to the spec because TopLink didn't support it, but perhaps EclipseLink does now (I have no idea).

Now, what _could_ we do? Well, in principle, I suppose, we could add support to the `Dialect`s for parsing and interpreting `columnDefinition`s, something sorta similar to what you proposed here earlier, but obviously it would need to be much more robust. And then we could build that into `Column` itself, so that your code:

```java
@Column(columnDefinition="integer")
Long version;
```

would be interpreted as if you had written:

```java
@JdbcTypeCode(INTEGER)
@Column(columnDefinition="integer")
Long version;
```

and:

```java
@Column(columnDefinition="decimal(10,2) not null")
Long version;
```

would be interpreted as if you had written:

```java
@JdbcTypeCode(DECIMAL)
@Column(columnDefinition="decimal(10,2) not null", 
        nullable=false, precision=10, scale=2)
Long version;
```

So how hard would it be to implement this? Well, not necessarily very hard I suppose. There are some type names which are a bit nontrivial to parse, e.g. `timestamp(6) with time zone`, `char(128) for bit data`, etc, and lots of dialect-specific variation in naming, e.g. `timestamptz`, `varchar2`, etc, but it's not _difficult_ work, just a bit of tedium involved in going through all the databases and compiling all the information.

I think it would be a very nice thing to have, to be honest, but I'm not motivated to lose a week of my life to this.

What I'm saying is: if you want to do this, do it _properly_, instead of hacking in something that sorta works in only the most trivial cases.

Reply:
I agree with you, but `columnDefinition` is JPA specification and `@JdbcTypeCode` is Hibernate specific that most people like me doesn't even know.

> I mean after this latest change you are now optimizing for the uncommon case, putting all the string manipulation up front. I have no idea if it's a significant performance hit.

Please review latest commit, It should avoid string manipulation for common case.

Reply:
What I'm saying is that I don't want to do *anything* here unless we do it properly. In particular I don't want to make this code more complex than it already is just to handle 1% of possible cases. That's just technical debt.

If you really believe that this is important, then it should be done according to something like the approach I've outlined.